mlflow/gateway/config.py

```diff
@@ -366,71 +477,175 @@ class RouteConfig(AliasedConfigModel):
-                f"'{route_name}' is invalid. Names cannot contain spaces or any non "
-                "alphanumeric characters other than hyphen and underscore."
-            )
-        return route_name
+    if IS_PYDANTIC_V2:
+        from pydantic import field_validator as _field_validator
+
+        @_field_validator("name", mode="before")
+        @classmethod
+        def validate_endpoint_name(cls, route_name):
+            if not is_valid_endpoint_name(route_name):
+                raise MlflowException.invalid_parameter_value(
+                    "The route name provided contains disallowed characters for a url endpoint. "
+                    f"'{route_name}' is invalid. Names cannot contain spaces or any non "
+                    "alphanumeric characters other than hyphen and underscore."
+                )
+            return route_name
+
+    else:
+        from pydantic import validator as _validator
+
+        @_validator("name", pre=True)
+        def validate_endpoint_name(cls, route_name):
+            if not is_valid_endpoint_name(route_name):
+                raise MlflowException.invalid_parameter_value(
+                    "The route name provided contains disallowed characters for a url endpoint. "
+                    f"'{route_name}' is invalid. Names cannot contain spaces or any non "
+                    "alphanumeric characters other than hyphen and underscore."
+                )
+            return route_name
+
+    if IS_PYDANTIC_V2:
+        from pydantic import field_validator as _field_validator
+
+        @_field_validator("model", mode="before")
+        @classmethod
+        def validate_model(cls, model):
+            if model:
+                model_instance = Model(**model)
+                if model_instance.provider in Provider.values() and model_instance.config is None:
+                    raise MlflowException.invalid_parameter_value(
+                        "A config must be supplied when setting a provider. The provider entry for "
+                        f"{model_instance.provider} is incorrect."
+                    )
+            return model
+
+    else:
+        from pydantic import validator as _validator
+
+        @_validator("model", pre=True)
+        def validate_model(cls, model):
+            if model:
+                model_instance = Model(**model)
+                if model_instance.provider in Provider.values() and model_instance.config is None:
+                    raise MlflowException.invalid_parameter_value(
+                        "A config must be supplied when setting a provider. The provider entry for "
+                        f"{model_instance.provider} is incorrect."
+                    )
+            return model
+
+    if IS_PYDANTIC_V2:
+        from pydantic import model_validator as _model_validator
+
+        @_model_validator(mode="after")
+        def validate_route_type_and_model_name(self) -> Self:
```

Comment:
```suggestion
        def validate_route_type_and_model_name(self):
```


The doc building job is also failing due to `ImportError: cannot import name 'Self' from 'typing' (/home/circleci/.pyenv/versions/3.8.16/lib/python3.8/typing.py)`. It seems the `Self` type is not supported by the our current version of typing extension.